Mathlib/LinearAlgebra/AffineSpace/AffineSubspace/Basic.lean

```diff
@@ -0,0 +1,822 @@
+open scoped Pointwise
+
+section
+
+variable (k : Type*) {V : Type*} {P : Type*} [Ring k] [AddCommGroup V] [Module k V]
+variable [AffineSpace V P]
+
+@[simp] lemma vectorSpan_vadd (s : Set P) (v : V) : vectorSpan k (v +ᵥ s) = vectorSpan k s := by
+  simp [vectorSpan]
+
+end
+
+
+namespace AffineSubspace
+
+variable (k : Type*) {V : Type*} (P : Type*) [Ring k] [AddCommGroup V] [Module k V]
+  [AffineSpace V P]
+
+variable {k P}
+
+
+/-- Given a point in an affine subspace, a result of subtracting that point on the right is in the
+direction if and only if the other point is in the subspace. -/
+theorem vsub_right_mem_direction_iff_mem {s : AffineSubspace k P} {p : P} (hp : p ∈ s) (p₂ : P) :
+    p₂ -ᵥ p ∈ s.direction ↔ p₂ ∈ s := by
+  rw [mem_direction_iff_eq_vsub_right hp]
+  simp
+
+/-- Given a point in an affine subspace, a result of subtracting that point on the left is in the
+direction if and only if the other point is in the subspace. -/
+theorem vsub_left_mem_direction_iff_mem {s : AffineSubspace k P} {p : P} (hp : p ∈ s) (p₂ : P) :
+    p -ᵥ p₂ ∈ s.direction ↔ p₂ ∈ s := by
+  rw [mem_direction_iff_eq_vsub_left hp]
+  simp
+
+-- See note [reducible non instances]
+/-- This is not an instance because it loops with `AddTorsor.nonempty`. -/
+abbrev toAddTorsor (s : AffineSubspace k P) [Nonempty s] : AddTorsor s.direction s where
+  vadd a b := ⟨(a : V) +ᵥ (b : P), vadd_mem_of_mem_direction a.2 b.2⟩
+  zero_vadd := fun a => by
+    ext
+    exact zero_vadd _ _
+  add_vadd a b c := by
+    ext
+    apply add_vadd
+  vsub a b := ⟨(a : P) -ᵥ (b : P), (vsub_left_mem_direction_iff_mem a.2 _).mpr b.2⟩
+  vsub_vadd' a b := by
+    ext
+    apply AddTorsor.vsub_vadd'
+  vadd_vsub' a b := by
+    ext
+    apply AddTorsor.vadd_vsub'
+
+attribute [local instance] toAddTorsor
+
+@[simp, norm_cast]
+theorem coe_vsub (s : AffineSubspace k P) [Nonempty s] (a b : s) : ↑(a -ᵥ b) = (a : P) -ᵥ (b : P) :=
+  rfl
+
+@[simp, norm_cast]
+theorem coe_vadd (s : AffineSubspace k P) [Nonempty s] (a : s.direction) (b : s) :
+    ↑(a +ᵥ b) = (a : V) +ᵥ (b : P) :=
+  rfl
+
+/-- Embedding of an affine subspace to the ambient space, as an affine map. -/
+protected def subtype (s : AffineSubspace k P) [Nonempty s] : s →ᵃ[k] P where
+  toFun := (↑)
+  linear := s.direction.subtype
+  map_vadd' _ _ := rfl
+
+@[simp]
+theorem subtype_linear (s : AffineSubspace k P) [Nonempty s] :
+    s.subtype.linear = s.direction.subtype := rfl
+
+@[simp]
+theorem subtype_apply {s : AffineSubspace k P} [Nonempty s] (p : s) : s.subtype p = p :=
+  rfl
+
+theorem subtype_injective (s : AffineSubspace k P) [Nonempty s] : Function.Injective s.subtype :=
+  Subtype.coe_injective
+
+@[simp]
+theorem coe_subtype (s : AffineSubspace k P) [Nonempty s] : (s.subtype : s → P) = ((↑) : s → P) :=
+  rfl
+
+@[deprecated (since := "2025-02-18")]
+alias coeSubtype := coe_subtype
+
+theorem injective_subtype (s : AffineSubspace k P) [Nonempty s] : Function.Injective s.subtype :=
+  Subtype.coe_injective
+
```

Comment:
This was removed on master in #22399; the declarations diff shows this PR is inadvertently adding it back.